crates/parcel/src/requests/asset_request.rs

```diff
@@ -0,0 +1,168 @@
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use parcel_core::cache::CacheRef;
+use parcel_core::plugin::AssetBuildEvent;
+use parcel_core::plugin::BuildProgressEvent;
+use parcel_core::plugin::InitialAsset;
+use parcel_core::plugin::ReporterEvent;
+use parcel_core::plugin::RunTransformContext;
+use parcel_core::plugin::TransformResult;
+use parcel_core::plugin::TransformationInput;
+use parcel_core::types::Asset;
+use parcel_core::types::AssetStats;
+use parcel_core::types::Dependency;
+use parcel_core::types::Environment;
+use parcel_core::types::FileType;
+use parcel_filesystem::FileSystemRef;
+
+use crate::plugins::Plugins;
+use crate::plugins::TransformerPipeline;
+use crate::request_tracker::{Request, RequestResult, RunRequestContext, RunRequestError};
+
+/// The AssetRequest runs transformer plugins on discovered Assets.
+/// - Decides which transformer pipeline to run from the input Asset type
+/// - Runs the pipeline in series, switching pipeline if the Asset type changes
+/// - Stores the final Asset source code in the cache, for access in packaging
+/// - Finally, returns the complete Asset and it's discovered Dependencies
+pub struct AssetRequest<'a> {
+  pub env: Arc<Environment>,
+  pub file_path: PathBuf,
+  pub code: Option<String>,
+  pub pipeline: Option<String>,
+  pub side_effects: bool,
+  // TODO: move the following to RunRequestContext
+  pub cache: CacheRef,
+  pub file_system: FileSystemRef,
+  pub plugins: Arc<Plugins<'a>>,
+}
+
+impl<'a> Hash for AssetRequest<'a> {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    // TODO: Just derive this once the contextual params are moved to RunRequestContext
+    self.file_path.hash(state);
+    self.code.hash(state);
+    self.pipeline.hash(state);
+    self.env.hash(state);
+    self.side_effects.hash(state);
+  }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct AssetResult {
+  pub asset: Asset,
+  pub dependencies: Vec<Dependency>,
+}
+
+impl<'a> Request<AssetResult> for AssetRequest<'a> {
+  fn run(
+    &self,
+    request_context: RunRequestContext<AssetResult>,
+  ) -> Result<RequestResult<AssetResult>, RunRequestError> {
+    request_context.report(ReporterEvent::BuildProgress(BuildProgressEvent::Building(
+      AssetBuildEvent {
+        // TODO: Should we try avoid a clone here?
+        file_path: self.file_path.clone(),
+      },
+    )));
+
+    let pipeline = self
+      .plugins
+      .transformers(&self.file_path, self.pipeline.as_deref())?;
+    let asset_type = FileType::from_extension(
+      self
+        .file_path
+        .extension()
+        .and_then(|s| s.to_str())
+        .unwrap_or(""),
+    );
+    let mut transform_ctx = RunTransformContext::new(self.file_system.clone());
+
+    let result = run_pipeline(
+      pipeline,
+      TransformationInput::InitialAsset(InitialAsset {
```

Comment:
Can the input to a transformer always be an asset? Ideally transformers shouldn't know or care that they are first in a pipeline

Reply:
The idea here is to abstract this away via methods, but definitely happy to iterate on this design.